### PR TITLE
Analysis design

### LINKED ISSUE #4: Discussion: the design of Analysis
(I promise that I'm done after this one for tonight :slightly_smiling_face: )

I would like to discuss the design of the `Analysis` abstraction.

---

Let me start with a distraction. In an ideal word, there would be a type class `MeetSemiLattice`, and `Analysis` would look more like

```haskell
class (Eq (Domain l), MeetSemiLattice (Domain l)) => Analysis l where
    type Domain l

    makeA :: ENode l -> EGraph l -> Domain l
    modifyA :: ClassId -> EGraph l -> EGraph l
```

There is a well-maintained [`lattices`](https://hackage.haskell.org/package/lattices). Unfortunately it doesn't have semi-lattices. And I don't think it's worth it to introduce a bespoke type class here, especially since the indirection creates some boilerplate. But I'm still a little sad

---

Ok, so, to the real point, here. Something that users are likely to want to do is to _combine analyses_. This is a rather standard thing in abstract interpretation, so let's take this example (it can be used for optimisation, _e.g._ to prove that certain numbers are not 0 so that we can do some mathematical simplification, or that some branches can't be reached so that we can eliminate them, etc… so it's rather relevant to the whole Egg thing, kind of).

So I can write, say, an interval analysis to approximate values of my number variable (this lets me answer questions like “can this expression evaluate to 0”). In my problem domain, parity happens to matter a lot (this is fictional, I personally have never needed parity), so I also write a parity analysis.

Then, using some kind of Cartesian product, I can combine these two analyses into one, and then run equality saturation using the combined analysis.

Some of my rewrite rules use the interval analysis exclusively, they can be developed independently of the parity analysis. And conversely the other rewrite rules use the parity analysis exclusively. (there may also be rules that use both at the same time, of course). The prerequisite, here, is that I can add new analyses later without changing my rules.

I contend that with the current tool, such a workflow is not supported.

The first impediment is that, due to how the type class is structured, there is a single possible analysis for a given `L`. This can be worked around by using `newtype` wrappers over `L`, like
```haskell
newtype Parity a = Parity (L a)
```
But then every rewrite rule will look like (borrowing an example from the README)
```haskell
pat (Parity (pat (Parity ("a" :*: "b")) :/: "c")) := pat (Parity ("a" :*: pat (Parity ("b" :/: "c"))))
```
this honestly doesn't look nice.

Then there are the members of the class
- `Domain` is easy: it's the product `(A, B)` of the domains of the two analyses
- `join` is easy: it's `join *** join`
- `make` is tricky because it's such a general function. But we usually don't need such a general function: generally, we just look at the analysis value of the children and we use that to compute our new analysis value. If we give `make` the type
  ```haskell
  make :: l a -> a
  ```
  then the Cartesian product of `make` is `\as -> (make (fmap fst as), make (fmap snd as))`. So at the price of a slightly more restricted function, we can combine `make`
- `modify` is quite a bit more difficult. First this is a very lightly specified function in the Egg paper. It's too general for combination to make sense. But even if we stick to the only example that I'm aware of – adding nodes – it is not clear what a combination would look like. Because the `modify` for `A` could add a node, which triggers the `modify` from `B` to add a node, which in turn triggers the `modify` for `A` to add another node, etc… There is no guarantee that this process would converge (and it's probably not particularly fast either). There I don't really have a way forward.

---

Anyway, this is the lesson that I've learnt from abstract interpretation: make small analyses and combine them. I think it's very relevant for Egg (hence `hegg`!). But, while I have some rough idea of what this would need to do, I don't really have a concrete suggestion that would improve the interface. I do understand the reason why the `Analysis` type class is the way it is, it is really not obvious to avoid.

Maybe you could have
```haskell
data EGraph analysis language
class Analysis analysis language
```
I'm worried this may yield an unhealthy amount of ambiguous variables, but it may be worth trying.

Though understanding how the `modify` function is necessary (as I'm writing this I'm questioning whether the `modify` function is really useful. At least in the case of constant folding, presented in some detail in the Egg paper, we could implement it as an analysis-guided rewrite rule).

To conclude: I don't really know what to think. I see this thread as an open discussion.

### PR DESCRIPTION
**Pull request by alt-romes**
Closes #4 

This isn't going too smoothly yet, it's a complicated endeavor.
We need definitely more use cases before pulling something like this, I think.

Things that probably can't remain as I've left it in this branch is `modifyA`, since I do think one might want to add recursive nodes, and that because just manually adding them to the class seems to make the test suite fail.
I thought about an alternative in which

```hs
modifyA :: EClass a l -> [ModifyActions]

data ModifyActions l = AddNode (ENode l) | AddExpr (Fix l) | DeleteNode (ENode l)
```
but unsure if that looks good enough, is flexible enough, etc...

So I mean to continue the discussion with this PR.

If anyone implementing analysis could show us their use case I'd be grateful 👍 

Thanks

### CODE REVIEW

**Comment by alt-romes** [comment]
Okay, I think I've mostly nailed this down @aspiwack, it'd be great to get some input.

I've done a big re-design of the analysis, and made it (I think) much more friendly, at the cost of some flexibility.

The driving change is the Analysis class definition.

First of all, we add a new parameter to the e-graph to indicate which analysis is to be used (as decided in #4)
Secondly, the methods were redesigned so knowledge of the e-graph isn't even necessary.

Both these together allow for combining together analysis. For example, an instance for `(Analysis a l, Analysis b l) => Analysis (a, b) l` is provided in `Data.Equality.Analysis.Instances`, and with it we can combine any two (law-abiding) analysis that depend on different domains into one combined analysis that depends on the product of the domains.

The analysis class now is (omitting some examples):

```hs
class Eq (Domain a l) => Analysis a (l :: Type -> Type) where

    -- | Domain of data stored in e-class according to e-class analysis
    type Domain a l

    -- | When a new e-node is added into a new, singleton e-class, construct a
    -- new value of the domain to be associated with the new e-class, typically
    -- (always?) by accessing the associated data of n's children
    --
    -- The argument is the e-node term populated with its children data
    makeA :: l (Domain a l) -> Domain a l

    -- | When e-classes c1 c2 are being merged into c, join d_c1 and
    -- d_c2 into a new value d_c to be associated with the new
    -- e-class c
    joinA :: Domain a l -> Domain a l -> Domain a l

    -- | Optionally modify the e-class c (based on d_c), typically by adding an
    -- e-node to c. Modify should be idempotent if no other changes occur to
    -- the e-class, i.e., modify(modify(c)) = modify(c)
    --
    -- The return value of the modify function is both the modified class and
    -- the expressions (in their fixed-point form) to add to this class. We
    -- can't manually add them because not only would it skip some of the
    -- internal steps of representing + merging, but also because it's
    -- impossible to add any expression with depth > 0 without access to the
    -- e-graph (since we must represent every sub-expression in the e-graph
    -- first).
    --
    -- That's why we must return the modified class and the expressions to add
    -- to this class.
    modifyA :: EClass a l -> (EClass a l, [Fix l])
```

Unsolved/undecided bits:

As you predicted, this does create a lot of ambiguous variables. I think that's okay; I would guess the user who understands data fix points enough to use this library is also comfortable with type applications.

However, I wrote down a possible alternative:
```hs
-- | TODO: If the domain is used in the class definition we would no longer have
-- ambiguous types neither the type family. That is, we'd have @Analysis domain
-- language@ instead of @Analysis analysis language@. Would that be better?
```
Haven't thought it through further than that.

Secondly, I didn't try to get the lambda tests to work in this model. This lambda test used by egg depends on the e-class id of the e-node children to manage free variables. Initially I thought we might be limited since we can no longer do this (depend on the class-ids for the analysis data), but then I thought we might just be able to emulate the behavior or even come up with another way to work the free variables.

Concluding,

I believe this analysis design is a much nicer cleaner/functional/compositional abstraction and I don't think yet it sacrifices too much. It's one good advantage the haskell implementation has over others!

I also think we can revise some of these changes if this more constrained flexibility turns out to be a problematic.

**Comment by aspiwack** [comment]
Can you explain why you have both the type parameter `a` and the associated family `Domain a l` in the type class? I would intuitively expect them to be one and the same (I'm guessing this would cut down on ambiguous variables too).

**Comment by alt-romes** [comment]
Yes! Almost. That's an alternative design I mentioned above. I'm starting to think it might be better altogether.

Right now, the `a` variable stands for the type of the analysis. For most cases it will be identical to the Domain a l, but what if we want to have two different analysis whose domain is the same (say `Maybe Int`).

Drawbacks: lots of ambiguous variables, often redundant type family instances

Pros: multiple instances using the same domain don't require newtype wrappers, we could use type level string literals to "name" the analysis.

I think changing it might be better; writing newtypes to wrap the same domain for use in different analysis doesn't sound bad

**Comment by alt-romes** [comment]
In writing it I see no reason not to change it.

What currently is is what came to be historically, and seems now unnecessarily complex

**Comment by alt-romes** [comment]
@aspiwack I've pushed the design using *domain* as a type variable.

* It allowed us to get rid of an import cycle (great!)
* It's overall simpler, the domain is a type variable, not a type instance of what basically was a type label
* We still have ambiguity at call sites like `equalitySaturation`, but there's no way around it (no domain is needed to call that)

I think it's pretty much ready to merge, and we'll have to understand the expressiveness problems when someone bumps into one.

**Review by aspiwack — COMMENTED** [review]
This looks pretty nice, in my opinion. Hopefully it will pan out in practice too!

File: src/Data/Equality/Graph.hs [hunk 1]
```diff
@@ -51,18 +54,28 @@ import Data.Equality.Analysis
...
 --     mempty = EGraph emptyUF mempty mempty mempty
 
+represent :: forall a l. (Analysis a l, Language l) => Fix l -> EGraph a l -> (ClassId, EGraph a l)
+represent = cata (flip $ \e -> uncurry add . first Node . (`runState` e) . traverse (gets >=> \(x,e') -> x <$ put e'))
+
+-- f :: l (EGraph a l -> (ClassId, EGraph a l)) -> EGraph a l -> (ClassId, EGraph a l)
+-- f l e = uncurry add $ first Node $ (`runState` e) $ traverse g l
+
+-- g :: (EGraph a l -> (ClassId, EGraph a l)) -> State (EGraph a l) ClassId
+-- g h = gets h >>= \(x,e) -> x <$ put e
```
Comment (alt-romes) [inline src/Data/Equality/Graph.hs:72]:
```suggestion
```

**Comment by alt-romes** [comment]
@aspiwack I've revised the comments, updated the tutorial and changelog and revived the lambda test on top of the analysis design.

The lambda test is now much much better than before because we do in fact make use of composition of analysis!
The lambda language analysis is really a combination of constant folding and free-variable tracking. It worked great and it freed me from the burden of combining them manually. I was also able to implement the analysis without loss of expressiveness. (Note that the lambda test still doesn't implement all tests from the paper, but it definitely does more so on this patch than it used to -- for completion I think we're still missing a redesign of rewrite conditions + dynamic rewrites)

Let's merge this!

File: src/Data/Equality/Analysis.hs [hunk 2]
```diff
@@ -23,41 +27,115 @@ References: https://arxiv.org/pdf/2004.03082.pdf
...
+-- otherwise
+instance forall l. Analysis () l where
+  makeA _ = ()
+  joinA = (<>)
+
+
+-- This instance is not necessarily well behaved for any two analysis, so care
+-- must be taken when using it.
+--
+-- A possible criteria is:
```
Comment (aspiwack) [inline src/Data/Equality/Analysis.hs:110]:
```suggestion
-- A possible criterion is:
```

File: src/Data/Equality/Analysis.hs [hunk 3]
```diff
@@ -23,41 +27,115 @@ References: https://arxiv.org/pdf/2004.03082.pdf
...
+--
+-- The @domain@ is the type of the domain of the e-class analysis, that is, the
+-- type of the data stored in an e-class according to this e-class analysis
+class Eq domain => Analysis domain (l :: Type -> Type) where
 
     -- | When a new e-node is added into a new, singleton e-class, construct a
-    -- new value of the domain to be associated with the new e-class, typically
-    -- by accessing the associated data of n's children
-    makeA :: ENode l -> EGraph l -> Domain l
+    -- new value of the domain to be associated with the new e-class, always by
```
Comment (aspiwack) [inline src/Data/Equality/Analysis.hs:43]:
```suggestion
    -- new value of the domain to be associated with the new e-class, by
```

**Review by aspiwack — COMMENTED** [review]
This is looking good.

**Comment by alt-romes** [comment]
Thanks for all the help @aspiwack, I'll release this soon in a v0.3

**Pull request merged**